test/built-ins/String/prototype/padEnd/observable-operations.js

```diff
@@ -0,0 +1,51 @@
+// Copyright (C) 2016 Jordan Harband. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-string.prototype.padend
+description: String#padEnd should perform observable operations in the correct order
+author: Jordan Harband
+---*/
+
+var log = "";
+
+function createPrimitiveObserver(name, string, value) {
+  return {
+    toString: function () {
+      log += '|toString:' + name;
+      return string;
+    },
+    valueOf: function () {
+      log += '|valueOf:' + name;
+      return value;
+    }
+  };
+};
+
+var receiver = createPrimitiveObserver('receiver', createPrimitiveObserver('toPrimitiveABC', {}, 'abc'));
+
+var fillString = createPrimitiveObserver('fillString', createPrimitiveObserver('toPrimitiveDEF', {}, 'def'));
+
+var maxLength = createPrimitiveObserver(
+    'maxLength',
+    createPrimitiveObserver('toPrimitiveMaxLengthToString', {}, {}),
+    createPrimitiveObserver('toPrimitiveMaxLengthValueOf', {}, 11)
+);
+
+var result = String.prototype.padEnd.call(receiver, fillString, maxLength);
+
+assert.sameValue(result, 'defdefdeabc');
```

Comment:
on https://tc39.github.io/ecma262/#sec-toprimitive

```
...
7. Return ? OrdinaryToPrimitive(input, hint). 
```

```
3. If hint is "string", then
  a. Let methodNames be « "toString", "valueOf" ».
...
5. For each name in methodNames in List order, do
  a. Let method be ? Get(O, name).
  b. If IsCallable(method) is true, then
    i. Let result be ? Call(method, O).
    ii. If Type(result) is not Object, return result.
...
```

So, step 5.b.ii will check the result for `receiver.toString` is an Object (`createPrimitiveObserver('toPrimitiveABC', {}, 'abc')`), and will continue the loop getting result from `receiver.valueOf`.